src/vm/i386/umthunkstub.S

```diff
@@ -139,30 +148,40 @@ LOCAL_LABEL(UMThunkStub_CopyStackArgs):
     // esi = src
     // edi = dest
     // ebx = scratch
-    lea     esi, [ebp + 0x08]
+    lea     esi, [ebp + 0x8]
 
-    // first [esi] goes to ecx, in LTR
+    sub     esp, eax
+    and     esp, -16    // align with 16 byte
+    lea     edi, [esp]
+
+    mov     edx, dword ptr [ebp - UMThunkStub_UMENTRYTHUNK_OFFSET]
+    mov     edx, dword ptr [edx + UMEntryThunk__m_pUMThunkMarshInfo]
+    mov     ebx, dword ptr [edx + UMThunkMarshInfo__m_ecxArgOffset]
+    cmp     ebx, -1
+    je      LOCAL_LABEL(CopyStack)
+    mov     ecx, dword ptr [esi + ebx]
     add     eax, -4
-    mov     ecx, dword ptr [esi]
     jz      LOCAL_LABEL(UMThunkStub_ArgumentsSetup)
 
-    // second [esi+04] goes to edx
+    mov     ebx, dword ptr [edx + UMThunkMarshInfo__m_edxArgOffset]
+    cmp     ebx, -1
+    je      LOCAL_LABEL(CopyStack)
+    mov     edx, dword ptr [esi + ebx]
     add     eax, -4
-    mov     edx, dword ptr [esi + 0x04]
     jz      LOCAL_LABEL(UMThunkStub_ArgumentsSetup)
 
-    sub     esp, eax
-    and     esp, -16    // align with 16 byte
-    lea     edi, [esp]
-
-LOCAL_LABEL(CopyLoop):
 
-    // copy rest of the arguments to [esp+08+n], in RTL
+LOCAL_LABEL(CopyStack):
```

Comment:
This doesn't look right. We need to copy all arguments that were not stored in the registers, in other words, the loop needs to skip copying the ones that were in registers to the new location of stack arguments. This loop does something different. EAX at the beginning is equal to the m_cbActualArgSize minus the size of args we pass in registers, so it is the size of the stack arguments. Let's call it K. The loop copies first K arguments to the target location on the stack. But that's not right, the arguments that are passed in registers can be at arbitrary offsets - that's why you have introduced the m_edxArgOffset and m_ecxArgOffset, right? So you could miss some arguments and some could get duplicated.
I think the simplest way to copy the arguments is to just loop through them from the first till the last one and compare the current offset with the m_edxArgOffset and m_ecxArgOffset. If the offset matches, then you load the source argument into a register. If it doesn't match, then you place it on the target stack.